gitutil.py

```diff
@@ -0,0 +1,137 @@
+
+    Parameters
+    ------
+    repo_path : str
+        Path to a directory containing an intialised git-repo with a submodule to update.
+    tree_ish : str
+        Valid tree-ish to use as base for creating the new commit. This will be used as parent for the commit created by this function.
+        Example: 'master' for the head of the master-branch.
+    submodule_path : str
+        Path (relative to `repo_path`) to the submodule. Must be immediately below the root (i.e. `repo_path`) of the repository.
+    commit_hash : str
+        The hash the submodule should point to in the created commit. This should be a valid commit hash in the submodule's repository.
+    author : str,
+        Will be set as author of the created commit
+    email : str
+        Will be set for the author of the created commit
+
+    Returns
+    ------
+    str
+        The hexadecimal SHA-1 hash of the created commit
+    '''
+    ensure_directory_exists(repo_path)
+    repo_path = os.path.abspath(repo_path)
+
+    ensure_not_empty(submodule_path)
+    if '/' in submodule_path:
+        fail('Specified submodule-path is not in the top-level of the repository: {s}'.format(s=submodule_path))
+
+    ensure_not_empty(tree_ish)
+    ensure_not_empty(commit_hash)
+    ensure_not_empty(author)
+    ensure_not_empty(email)
+
+    repo = git.Repo(repo_path)
+
+    # Fetch tree-ish from git.repo and create mktree-parseable string-representation of it.
+    # Since submodule-objects contain the SHA of the to-be-checked-out version of the
+    # target repository, we replace that during the string building.
+    tree = repo.tree(tree_ish)
+    tree_representation = '\n'.join(_serialise_and_patch_tree(tree, submodule_path, commit_hash))
+
+    # Pass the serialised tree to git mk-tree using GitPython. We cannot do this in GitPython
+    # directly as it does not support arbitrary tree manipulation.
+    auto_interrupt = repo.git.mktree(istream = subprocess.PIPE, as_process=True)
+    process = auto_interrupt.proc
+    stdout, _= process.communicate(input=tree_representation.encode())
+
+    # returned string is byte-encoded and newline-terminated
+    new_sha = stdout.decode('utf-8').strip()
+
+    # Create a new commit in the repo's object database from the newly created tree.
+    actor = git.Actor(author, email)
+    parent_commit = repo.commit(tree_ish)
+    commit = git.Commit.create_from_tree(
+      repo = repo,
+      tree = new_sha,
+      parent_commits = [parent_commit],
+      message='Upgrade submodule {s} to commit {c}'.format(
+          s=submodule_path,
+          c=commit_hash,
+      ),
+      author=actor,
+      committer=actor,
+    )
+
+    return commit.hexsha
+
+
+def _serialise_and_patch_tree(
+    tree: git.Tree,
+    submodule_path: str,
+    commit_hash: str,
+):
+    '''Create a generator that yields the string-representation of the passed tree as understood by
```

Comment:
the fact that a generator is returned is correct; but this is not the main aspect.

-> this function accepts as input a tree and a submodule path with a commit and returns a modified serialised tree representation, where the entry pointing to the submodule is altered and all other entries are left unmodified.

it is also sufficient to state that the result is returned as an iterable (although it _might_ be more adequate to return the full string, as the single lines are probably of no use to anyone)